src/platform/nrfconnect/OTAMultiImageDownloader.cpp

```diff
@@ -0,0 +1,181 @@
+
+int OTAMultiImageDownloader::Init(const device * flash, size_t offset, size_t size)
+{
+    mBuffer = (uint8_t *) k_malloc(kBufferSize);
+    return stream_flash_init(&mStream, flash, mBuffer, kBufferSize, offset, size, nullptr);
+}
+
+int OTAMultiImageDownloader::Apply(const device * flash)
+{
+    // Reinit stream with different flash partition
+    int err = Init(flash, mTargetOffset, mTargetImgSize);
+    if (err != 0)
+    {
+        ChipLogError(SoftwareUpdate, "Cannot reinitialize flash stream");
+        return -EBUSY;
+    }
+    uint8_t * buffer = (uint8_t *) k_malloc(kBufferSize);
+    if (!buffer)
+    {
+        return -ENOMEM;
+    }
+
+    for (size_t offset = 0; offset < mImgSize; offset += kBufferSize)
+    {
+        err = flash_read(flash, mOffset + offset, buffer, kBufferSize);
+        if (err != 0)
+        {
+            ChipLogError(SoftwareUpdate, "Failed to read data chunk from flash");
+            return -EIO;
+        }
+        err = Write(buffer, kBufferSize);
+        if (err != 0)
+        {
+            ChipLogError(SoftwareUpdate, "Failed to write data chunk to flash");
+            return -EIO;
+        }
+    }
+    k_free(buffer);
+
+    return Finalize();
+}
+
+int OTAMultiImageDownloader::Write(const uint8_t * chunk, size_t chunkSize)
+{
+    return stream_flash_buffered_write(&mStream, chunk, chunkSize, false);
+}
+
+int OTAMultiImageDownloader::Finalize()
+{
+    int err = stream_flash_buffered_write(&mStream, nullptr, 0, true);
+    k_free(mBuffer);
+    return err;
+}
+
+// Implementation of stateless OTAMultiImageDownloaders class
+
+int OTAMultiImageDownloaders::Open(int id, size_t size)
+{
+    int err{ -ENXIO };
+    if (CONFIG_APPLICATION_IDX == RunningVariant::THREAD && CONFIG_APPLICATION_OTHER_IDX == RunningVariant::WIFI)
+    {
+        if (id == CurrentImageID::APP_1_CORE_APP || id == CurrentImageID::APP_1_CORE_NET)
+        {
+            err = dfu_target_init(DFU_TARGET_IMAGE_TYPE_MCUBOOT, id, size, nullptr);
+            if (err)
+            {
+                ChipLogError(SoftwareUpdate, "Cannot initialize DFU target: %d", err);
+                return err;
+            }
+        }
+    }
+    else if (CONFIG_APPLICATION_IDX == RunningVariant::WIFI && CONFIG_APPLICATION_OTHER_IDX == RunningVariant::THREAD)
+    {
+        if (id == CurrentImageID::APP_2_CORE_APP || id == CurrentImageID::APP_2_CORE_NET)
+        {
+            err = dfu_target_init(DFU_TARGET_IMAGE_TYPE_MCUBOOT, id, size, nullptr);
+            if (err)
+            {
+                ChipLogError(SoftwareUpdate, "Cannot initialize DFU target: %d", err);
+                return err;
+            }
+        }
+    }
+
+    OTAMultiImageDownloaders::SetCurrentImageId(id);
+    return OTAMultiImageDownloaders::GetCurrentImage()->Init(sExtFlash, size);
+}
+
+int OTAMultiImageDownloaders::Write(const uint8_t * chunk, size_t chunk_size)
+{
+    if (OTAMultiImageDownloaders::CurrentImageId() <= CurrentImageID::APP_1_CORE_NET)
```

Comment:
I struggle to understand this condition. What if we're currently running a Wi-Fi variant?

Reply:
Yep, look like a bug. Will fix it.